staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -369,3 +325,56 @@ func WriteWatchCacheSizes(watchCacheSizes map[schema.GroupResource]int) ([]strin
 	}
 	return cacheSizes, nil
 }
+
+var _ serverstorage.StorageFactory = &StorageConfigFactory{}
+
+type StorageConfigFactory struct {
+	StorageConfig storagebackend.Config
+}
+
+func (s *StorageConfigFactory) NewConfig(resource schema.GroupResource) (*storagebackend.ConfigForResource, error) {
+	return s.StorageConfig.ForResource(resource), nil
+}
+
+func (s *StorageConfigFactory) ResourcePrefix(resource schema.GroupResource) string {
+	return resource.Group + "/" + resource.Resource
+}
+
+func (s *StorageConfigFactory) Backends() []serverstorage.Backend {
+	// nothing should ever call this method but we still provide a functional implementation
+	return (&serverstorage.DefaultStorageFactory{StorageConfig: s.StorageConfig}).Backends()
+}
+
+var _ serverstorage.StorageFactory = &transformerStorageFactory{}
+
+type transformerStorageFactory struct {
+	delegate             serverstorage.StorageFactory
+	transformerOverrides map[schema.GroupResource]value.Transformer
```

Comment:
I'm not sure we want a fixed map of groupresource → transformer overrides plumbed/locked in place... don't we need a dynamic lookup?

Reply:
We only have static lookup today, and do not need dynamic lookup until we have #112050 so I have deferred changing the structure here.

Reply:
Note that with this change and the [POC I linked above](https://github.com/enj/kubernetes/compare/enj/r/kms_load_once_v2...enj:kubernetes:enj/r/kms_load_once_v3?w=1), all of the dynamic lookup would be contained within this `transformerStorageFactory` struct for #112050 (along with the file watcher look to keep the mutable transformers updated).

Reply:
Once #104165 is fixed, we will also need the dynamic lookup to enable #111977